[USM] Support HTTP2 request cases

amitslavin opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.

-->
### What does this PR do?

This PR supports `Literal Never Indexed` and `Literal Without Indexing` for HTTP/2 decoding.
While parsing the literal cases, there are two additional cases that we have not yet supported:

- Literal Header Field without Indexing, where the top four bits are 0000 (referenced in: [RFC7541 Section 6.2.2](https://httpwg.org/specs/rfc7541.html#rfc.section.6.2.2))

- Literal Header Field never Indexed, where the top four bits are 0001 (referenced in: [RFC7541 Section 6.2.3](https://httpwg.org/specs/rfc7541.html#rfc.section.6.2.3))

These cases were identified during the TLS support development and are pertinent to the following [commit](https://github.com/DataDog/datadog-agent/pull/21619/commits/64b801ede094d155bd8c6f627facb98b31ad8f9b).

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

Supports decoding all possible HTTP/2 cases to handle various types of traffic.

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided. Except if the `qa/skip-qa` label, with required either `qa/done` or `qa/no-code-change` labels, are applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

amitslavin committed:
add support for Literal Header Field without Indexing and Literal Header Field never Indexed

guyarb commented on pkg/network/ebpf/c/protocols/http2/decoding.h:
```diff
@@ -248,11 +255,18 @@ static __always_inline void process_headers(struct __sk_buff *skb, dynamic_table
+                dynamic_value.is_huffman_encoded = current_header->is_huffman_encoded;
+                bpf_map_update_elem(&http2_dynamic_table, dynamic_index, &dynamic_value, BPF_ANY);
+            }
             bpf_map_update_elem(&http2_dynamic_table, dynamic_index, &dynamic_value, BPF_ANY);
```
should be deleted as you added it in the `if (current_header->type == kNewDynamicHeader)`

amitslavin replied:
removed

guyarb commented on pkg/network/ebpf/c/protocols/http2/decoding.h:
```diff
@@ -248,11 +255,18 @@ static __always_inline void process_headers(struct __sk_buff *skb, dynamic_table
             bpf_memcpy(current_stream->request_path, dynamic_value->buffer, HTTP2_MAX_PATH_LEN);
         } else {
+            // We're in new dynamic header or new dynamic header not indexed states.
             dynamic_value.string_len = current_header->new_dynamic_value_size;
             dynamic_value.is_huffman_encoded = current_header->is_huffman_encoded;
```
should be deleted as you added it in the `if (current_header->type == kNewDynamicHeader)`

guyarb commented on pkg/network/usm/monitor_test.go:
```diff
@@ -978,16 +978,31 @@ var (
-func createRawRequestFrame(streamID int) []byte {
-	binary.BigEndian.PutUint16(request[7:9], uint16(streamID))
-	return request
+const (
+	caseWithoutIndexing = 1
+	caseNeverIndexed    = 2
+)
```
```suggestion
type case int
const (
	caseWithoutIndexing case = iota
	caseNeverIndexed    case
)
```

guyarb commented on pkg/network/usm/monitor_test.go:
```diff
@@ -978,16 +978,31 @@ var (
+)
+
+// createHTTP2RequestByCase generates an HTTP/2 request based on the specified case.
+func createHTTP2RequestByCase(requestType int) []byte {
```
```suggestion
func createHTTP2RequestByCase(requestType case) []byte {
```

guyarb commented on pkg/network/usm/monitor_test.go:
```diff
@@ -978,16 +978,31 @@ var (
+// createHTTP2RequestByCase generates an HTTP/2 request based on the specified case.
+func createHTTP2RequestByCase(requestType int) []byte {
+	switch requestType {
+	// http2 request to path /aaa without indexing (same as request but with different index changed from 0x45 to 0x05)
+	case caseWithoutIndexing:
+		request[22] = 0x05
+		return request
+	// http2 request to path /aaa which never indexed (same as request but with different index changed from 0x45 to 0x15)
+	case caseNeverIndexed:
+		request[22] = 0x15
+		return request
```
maybe it is better to write the binary mode

guyarb replied:
redundant

guyarb commented on pkg/network/usm/monitor_test.go:
```diff
@@ -978,16 +978,31 @@ var (
+	// http2 request to path /aaa without indexing (same as request but with different index changed from 0x45 to 0x05)
+	case caseWithoutIndexing:
+		request[22] = 0x05
+		return request
```
redundant

guyarb commented on pkg/network/usm/monitor_test.go:
```diff
@@ -978,16 +978,31 @@ var (
+	case caseNeverIndexed:
+		request[22] = 0x15
+		return request
+	default:
+		return request
+	}
```
```suggestion
	default:
	}
	return request
```

guyarb reviewed (changes requested):
reviewed

amitslavin committed:
fixed cr notes

guyarb commented on pkg/network/usm/monitor_test.go:
```diff
@@ -964,39 +964,38 @@ var (
-func createHTTP2RequestByCase(requestType int) []byte {
+func createHTTP2RequestByCase(requestType literalType) []byte {
+	// http2 request to path /aaa with index 5
+	request := []byte{
+		0x00, 0x00, 0x37, 0x01, 0x04, 0x00, 0x00, 0x00, 0x03, 0x41, 0x8a, 0x08, 0x9d, 0x5c, 0x0b, 0x81,
+		0x70, 0xdc, 0x78, 0x0f, 0x0b, 0x83, 0x45, 0x83, 0x60, 0x63, 0x1f, 0x86, 0x5f, 0x8b, 0x1d, 0x75,
+		0xd0, 0x62, 0x0d, 0x26, 0x3d, 0x4c, 0x74, 0x41, 0xea, 0x5c, 0x01, 0x34, 0x50, 0x83, 0x9b, 0xd9,
+		0xab, 0x7a, 0x8d, 0xc4, 0x75, 0xa7, 0x4a, 0x6b, 0x58, 0x94, 0x18, 0xb5, 0x25, 0x81, 0x2e, 0x0f,
+	}
```
```suggestion
	result := make([]byte, len(request))
	copy(result, request)
```
and now work on the `result`

guyarb replied:
you're missing the copy

guyarb commented on pkg/network/usm/monitor_test.go:
```diff
@@ -1025,7 +1024,7 @@ func (s *USMHTTP2Suite) TestRawTraffic() {
 			numberOfSettingFrames: 100,
 			numberOfRequestFrames: 1,
-			requestBuffer:         request,
+			requestBuffer:         createHTTP2RequestByCase(caseDefault),
```
Why? you shouldn't be touching the unrelated tests

guyarb commented on pkg/network/usm/monitor_test.go:
```diff
@@ -964,30 +964,44 @@ var (
 		0x00, 0x00, 0x00, 0x04, 0x00, 0x00, 0x00, 0x00, 0x00,
 	}
 
-	// http2 request to path /aaa
-	request = []byte{
-		0x00, 0x00, 0x37, 0x01, 0x04, 0x00, 0x00, 0x00, 0x03, 0x41, 0x8a, 0x08, 0x9d, 0x5c, 0x0b, 0x81,
-		0x70, 0xdc, 0x78, 0x0f, 0x0b, 0x83, 0x45, 0x83, 0x60, 0x63, 0x1f, 0x86, 0x5f, 0x8b, 0x1d, 0x75,
-		0xd0, 0x62, 0x0d, 0x26, 0x3d, 0x4c, 0x74, 0x41, 0xea, 0x5c, 0x01, 0x34, 0x50, 0x83, 0x9b, 0xd9,
-		0xab, 0x7a, 0x8d, 0xc4, 0x75, 0xa7, 0x4a, 0x6b, 0x58, 0x94, 0x18, 0xb5, 0x25, 0x81, 0x2e, 0x0f,
-	}
-
```
restore

guyarb commented on pkg/network/usm/monitor_test.go:
```diff
@@ -964,30 +964,44 @@ var (
+const (
+	caseWithoutIndexing literalType = iota
+	caseNeverIndexed
+	caseDefault
```
redundant, remove

amitslavin committed:
fixed cr

guyarb commented on pkg/network/usm/monitor_test.go:
```diff
@@ -978,16 +978,31 @@ var (
+)
+
+// createHTTP2RequestByCase generates an HTTP/2 request based on the specified case.
+func createHTTP2RequestByCase(requestType literalType) []byte {
+	result := make([]byte, len(request))
+	switch requestType {
+	// http2 request to path /aaa without indexing (same as request but with different index changed from 0x45 to 0x05)
+	case caseWithoutIndexing:
+		result[22] = 0x05
+	// http2 request to path /aaa which never indexed (same as request but with different index changed from 0x45 to 0x15)
+	case caseNeverIndexed:
+		result[22] = 0x15
+	}
+	return result
```
https://github.com/DataDog/datadog-agent/pull/21913#discussion_r1444560086

amitslavin committed:
fixed test

guyarb commented:
please elaborate in the PR description and add the commit in grpc-tls PR that you took, and about the new 2 use cases we support now

amitslavin committed:
Merge remote-tracking branch 'origin/main' into amit.slavin/USMON-679-support-literal-use-cases

amitslavin committed:
Merge remote-tracking branch 'origin/main' into amit.slavin/USMON-679-support-literal-use-cases

amitslavin committed:
test

guyarb committed:
usm: http2: tests: Eliminate panic in tests

In a case where we have a bug in USM and it fails to start, we would run the test, but only at its end
when we tried to dump maps, we would have caused a nil-deref wihle trying to access usm pointer.

Also removes redundant Pause method in TestLargeBodiesGRPCScenarios

amitslavin committed:
Merge remote-tracking branch 'origin/main' into amit.slavin/USMON-679-support-literal-use-cases

amitslavin committed:
fix verifier issue

amitslavin committed:
try to fix issue

amitslavin committed:
Merge remote-tracking branch 'origin/main' into amit.slavin/USMON-679-support-literal-use-cases

# Conflicts:
#	pkg/network/usm/monitor_test.go

amitslavin committed:
support new tests

guyarb commented on pkg/network/protocols/http2/raw_traffic.go:
```diff
@@ -10,14 +10,17 @@ package http2
 import (
 	"bytes"
 	"fmt"
-
```
revert

guyarb commented on pkg/network/protocols/http2/raw_traffic.go:
```diff
@@ -10,14 +10,17 @@ package http2
 func NewHeadersFrameMessage(headerFields []hpack.HeaderField) ([]byte, error) {
 	var buf bytes.Buffer
 	enc := hpack.NewEncoder(&buf)
+	// we set the max dynamic table size to 100 to be able to test different cases of literal header parsing.
+	enc.SetMaxDynamicTableSizeLimit(DynamicTableSize)
```
I'm not sure this is right to do it everytime - maybe an optional parameter instead or split to functions

guyarb commented on pkg/network/usm/monitor_test.go:
```diff
@@ -953,26 +953,44 @@ func writeInput(c net.Conn, input []byte, timeout time.Duration) error {
+func generateTestHeaderFields(pathNeverIndexed, withoutIndexing bool) []hpack.HeaderField {
+	pathHeaderField := hpack.HeaderField{Name: ":path", Value: "/aaa", Sensitive: false}
+
+	// If the path is sensitive, we want to make sure that the path is longer than dynamicTableSize characters.
```
The documentation does not make sense
if the path is sensitive, we want a longer path?

guyarb commented on pkg/network/usm/monitor_test.go:
```diff
@@ -953,26 +953,44 @@ func writeInput(c net.Conn, input []byte, timeout time.Duration) error {
+	// The indexing is determined by the dynamic table size (which we set to dynamicTableSize) and the size of the path.
+	// Therefore, we want to make sure that the path is longer than 100 characters so that the path will not be indexed.
+	// ref: https://github.com/golang/net/blob/07e05fd6e95ab445ebe48840c81a027dbace3b8e/http2/hpack/encode.go#L140
+	if withoutIndexing {
+		pathHeaderField = hpack.HeaderField{Name: ":path", Value: "/" + strings.Repeat("a", usmhttp2.DynamicTableSize), Sensitive: true}
+	}
+
+	// If the path is sensitive, we are in a case where a path is never indexed
+	if pathNeverIndexed {
+		pathHeaderField = hpack.HeaderField{Name: ":path", Value: "/aaa", Sensitive: true}
+	}
```
you're allowing passing both options as "true", which will not work for you

guyarb commented on pkg/network/usm/monitor_test.go:
```diff
@@ -953,26 +953,44 @@ func writeInput(c net.Conn, input []byte, timeout time.Duration) error {
-	{Name: "content-length", Value: "4"},
-	{Name: "accept-encoding", Value: "gzip"},
-	{Name: "user-agent", Value: "Go-http-client/2.0"},
+// generateTestHeaderFields generates a set of header fields that will be used for the tests.
+func generateTestHeaderFields(pathNeverIndexed, withoutIndexing bool) []hpack.HeaderField {
```
You can add wrappers 

```suggestion
// generateTestHeaderFields generates a set of header fields that will be used for the tests.
func headersWithNeverIndexedPath() []hpack.HeaderField { return generateTestHeaderFields(...) }
func headersWithoutIndexingPath() []hpack.HeaderField { return generateTestHeaderFields(...) }
func defaultHeaders() []hpack.HeaderField { return generateTestHeaderFields(...) }
```

guyarb replied:
that way you dont need to pass `false, false` for most of the instances or `true, false` or `false, true`

amitslavin committed:
fix cr notes

amitslavin committed:
make the DynamicTableSize optional

amitslavin committed:
make the DynamicTableSize optional in other functions

amitslavin committed:
typo
